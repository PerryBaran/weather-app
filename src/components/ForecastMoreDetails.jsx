import React from "react";
import "../styles/ForecastDetails.css";
import PropTypes from "prop-types";
import { v4 as uuid } from "uuid";
import ForecastSummaries from "./ForecastSummaries";
import ForecastBreakdown from "./ForecastBreakdown";
import { calcMean, calcMax, calcMin } from "../helpers/calculateValues";
import { dateString } from "../helpers/dateTime";

function ForecastMoreDetails({ forecasts }) {
  const getTriHourlyForecasts = (triHourlyForecasts) => {
    const NUMBER_OF_ENTRIES = 24 / 3;
    const filteredForecastsArray = [];

    for (let i = 0; i < NUMBER_OF_ENTRIES; i += 1) {
      let time = `${i * 3}:00`;
      if (time.length === 4) {
        time = `0${time}`;
      }

      const triHourlyForecast = triHourlyForecasts.find(
        (forecast) => forecast.dt_txt.slice(11, 16) === time
      );

      if (triHourlyForecast) {
        const {
          main: { temp },
          weather: [{ description, id }],
        } = triHourlyForecast;

        filteredForecastsArray.push({
          dateOrTime: time,
          icon: id.toString(),
          key: uuid(),
          temp: temp.toFixed(2),
          weather: description,
        });
      } else {
        filteredForecastsArray.push({
          dateOrTime: time,
          icon: null,
          key: uuid(),
          temp: null,
          weather: null,
        });
      }
    }
    return filteredForecastsArray;
  };

  const triHourlyForecasts = getTriHourlyForecasts(forecasts);

  const day = dateString(forecasts[0].dt);
  const icon = forecasts[0].weather[0].id.toString();
  const temp = calcMean(forecasts, "main", "temp", 2);
  const minTemp = calcMin(forecasts, "main", "temp_max", 2);
  const maxTemp = calcMax(forecasts, "main", "temp_max", 2);
  const humidity = calcMean(forecasts, "main", "humidity", 0);
  const windSpeed = calcMean(forecasts, "wind", "speed", 2);
  const weather = forecasts[0].weather[0].description;

  return (
    <div className="forecast-details">
      <h2>{day}</h2>
      <ForecastSummaries forecasts={triHourlyForecasts} />
      <ForecastBreakdown
        icon={icon}
        temp={temp}
        minTemp={minTemp}
        maxTemp={maxTemp}
        humidity={humidity}
        windSpeed={windSpeed}
        weather={weather}
      />
    </div>
  );
}

ForecastMoreDetails.propTypes = {
  forecasts: PropTypes.arrayOf(
    PropTypes.shape({
      dt: PropTypes.number,
      dt_txt: PropTypes.string,
      main: PropTypes.shape({
        humidity: PropTypes.number,
        temp: PropTypes.number,
        temp_max: PropTypes.number,
        temp_min: PropTypes.number,
      }),
      weather: PropTypes.arrayOf(
        PropTypes.shape({
          description: PropTypes.string,
          id: PropTypes.number,
        })
      ),
      wind: PropTypes.shape({
        speed: PropTypes.number,
      }),
    })
  ).isRequired,
};

export default ForecastMoreDetails;
