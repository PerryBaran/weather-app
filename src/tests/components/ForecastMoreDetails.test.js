import React from "react";
import { render } from "@testing-library/react";
import ForecastMoreDetails from "../../components/ForecastMoreDetails";

xdescribe("ForecastMoreDetails", () => {
  const validProps = {
    forecasts: [
      {
        time: "now",
        icon: "200",
        temp: 20,
        maxTemp: 21,
        minTemp: 19,
        humdidity: 69,
        weather: "very cold",
        windSpeed: 4,
      },
      {
        time: "later",
        icon: "400",
        temp: 10,
        maxTemp: 11,
        minTemp: 9,
        humdidity: 96,
        weather: "more cold",
        windSpeed: 88,
      },
    ],
  };

  describe("snapshot", () => {
    it("renders correctly", () => {
      const { asFragment } = render(
        <ForecastMoreDetails forecasts={validProps.forecasts} />
      );

      expect(asFragment()).toMatchSnapshot();
    });
  });
});
