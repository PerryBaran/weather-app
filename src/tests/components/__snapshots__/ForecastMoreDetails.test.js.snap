// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForecastMoreDetails snapshot renders correctly 1`] = `
<DocumentFragment>
  <div
    class="forecast-details"
  >
    <h2>
      Thu Jan 01 1970
    </h2>
    <div
      class="forecast-summaries"
    >
      <div
        class="forecast-summary"
        data-testid="forecast-summary"
      >
        <p
          class="forecast-summary__dateOrTime"
        >
          00:00
        </p>
        <div
          class="forecast-summary__icon"
          data-testid="forecast-icon"
        >
          No Data
        </div>
        <p
          class="forecast-summary__description"
        />
        <p
          class="forecast-summary__temp"
        />
      </div>
      <div
        class="forecast-summary"
        data-testid="forecast-summary"
      >
        <p
          class="forecast-summary__dateOrTime"
        >
          03:00
        </p>
        <div
          class="forecast-summary__icon"
          data-testid="forecast-icon"
        >
          No Data
        </div>
        <p
          class="forecast-summary__description"
        />
        <p
          class="forecast-summary__temp"
        />
      </div>
      <div
        class="forecast-summary"
        data-testid="forecast-summary"
      >
        <p
          class="forecast-summary__dateOrTime"
        >
          06:00
        </p>
        <div
          class="forecast-summary__icon"
          data-testid="forecast-icon"
        >
          No Data
        </div>
        <p
          class="forecast-summary__description"
        />
        <p
          class="forecast-summary__temp"
        />
      </div>
      <div
        class="forecast-summary"
        data-testid="forecast-summary"
      >
        <p
          class="forecast-summary__dateOrTime"
        >
          09:00
        </p>
        <div
          class="forecast-summary__icon"
          data-testid="forecast-icon"
        >
          No Data
        </div>
        <p
          class="forecast-summary__description"
        />
        <p
          class="forecast-summary__temp"
        />
      </div>
      <div
        class="forecast-summary"
        data-testid="forecast-summary"
      >
        <p
          class="forecast-summary__dateOrTime"
        >
          12:00
        </p>
        <div
          class="forecast-summary__icon"
          data-testid="forecast-icon"
        >
          No Data
        </div>
        <p
          class="forecast-summary__description"
        />
        <p
          class="forecast-summary__temp"
        />
      </div>
      <div
        class="forecast-summary"
        data-testid="forecast-summary"
      >
        <p
          class="forecast-summary__dateOrTime"
        >
          15:00
        </p>
        <div
          class="forecast-summary__icon"
          data-testid="forecast-icon"
        >
          No Data
        </div>
        <p
          class="forecast-summary__description"
        />
        <p
          class="forecast-summary__temp"
        />
      </div>
      <div
        class="forecast-summary"
        data-testid="forecast-summary"
      >
        <p
          class="forecast-summary__dateOrTime"
        >
          18:00
        </p>
        <div
          class="forecast-summary__icon"
          data-testid="forecast-icon"
        >
          No Data
        </div>
        <p
          class="forecast-summary__description"
        />
        <p
          class="forecast-summary__temp"
        />
      </div>
      <div
        class="forecast-summary"
        data-testid="forecast-summary"
      >
        <p
          class="forecast-summary__dateOrTime"
        >
          21:00
        </p>
        <div
          class="forecast-summary__icon"
          data-testid="forecast-icon"
        >
          No Data
        </div>
        <p
          class="forecast-summary__description"
        />
        <p
          class="forecast-summary__temp"
        />
      </div>
    </div>
    <div
      class="forecast-breakdown"
    >
      <div
        class="forecast-breakdown__icon"
        data-testid="forecast-icon"
      >
        <i
          class="wi wi-day-sunny"
          name="owm"
        />
      </div>
      <p
        class="forecast-breakdown__weather"
      >
        desc 1
      </p>
      <p
        class="forecast-breakdown__temp"
      >
        <span>
          Temperture: 
        </span>
        20.50°C
      </p>
      <p
        class="forecast-breakdown__min-temp"
      >
        <span>
          Minimum Temperature: 
        </span>
        22.00°C
      </p>
      <p
        class="forecast-breakdown__max-temp"
      >
        <span>
          Maximum Temperature: 
        </span>
        23.00°C
      </p>
      <p
        class="forecast-breakdown__humidity"
      >
        <span>
          Humidity: 
        </span>
        11%
      </p>
      <p
        class="forecast-breakdown__wind-speed"
      >
        <span>
          Wind Speed: 
        </span>
        13.50m/s
      </p>
    </div>
  </div>
</DocumentFragment>
`;
